### PR TITLE
[debugger] Access invalid memory address using PointerValue Command.

### LINKED ISSUE #18191: Mono crashes on attemp to evaluate 'ref' returning method with debugger
## Steps to Reproduce
I've tried it with Rider 2019.3, but I guess it's also reproducible in other IDEs, because Rider does nothing special except the method invocation via SDB protocol

1. Create simple project with following code

```csharp
    internal class Program
    {
        private static int StaticF = 200;
        
        public static void Main(string[] args)
        { // set breakpoint here and evaluate 'RefReturn()'
        }
        
        public static ref int RefReturn()
        {
            return ref StaticF;
        }

    }
```

2.  Set a breakpoint on the start of Main()
3.  Start debugging and try to evaluate expression 'RefReturn()'

### Current Behavior
Mono crashes after a couple of seconds
`Process finished with exit code -1,073,741,819`

### Expected Behavior

The expression should be evaluated correctly

## On which platforms did you notice this

[ ] macOS (not tested)
[ ] Linux (not tested)
[ x] Windows

**Version Used**:
Mono JIT compiler version 6.6.0 (Visual Studio built mono)

### PR DESCRIPTION
**Pull request by thaystg**
Creating a fork process to access address that came from IDE using PointerValue. The IDE can send an invalid address and it was crashing mono.

Fixes #18191

### CODE REVIEW

**Comment by thaystg** [comment]
> 1. Interesting approach
> 2. How often does the debugger send these pointer commands? Does it only happen for user expression evaluation, or also for every local variable in a frame  or other things like that that can happen on every step?
> 3. Usual convention is `exit(0)` for success, and 1 for failure

The debugger send these pointer commands on every step if there is a pointer variable like Int32*, I assumed that this is not VERY used, do you disagree?

**Comment by lambdageek** [comment]
This is also related to https://github.com/mono/mono/issues/15612

File: mono/mini/debugger-agent.c [hunk 1]
```diff
@@ -9524,7 +9530,20 @@ pointer_commands (int command, guint8 *p, guint8 *end, Buffer *buf)
...
-		buffer_add_value (buf, m_class_get_byval_arg (m_class_get_element_class (klass)), (gpointer)addr, domain);
+		pid = fork ();
+		if (pid == 0) {
+			mono_runtime_cleanup_handlers ();
+			buffer_add_value (buf, m_class_get_byval_arg (m_class_get_element_class (klass)), (gpointer)addr, domain);
+			exit(0);
+		} else {
+			/* Parent */
+			waitpid (pid, &exit_status, 0);
+			if (!WIFEXITED (exit_status))
```
Comment (lambdageek) [inline mono/mini/debugger-agent.c]:
```suggestion
			if (!WIFEXITED (exit_status) && WEXITSTATUS (exit_status) == 0)
```

File: mono/mini/debugger-agent.c [hunk 2]
```diff
@@ -9509,17 +9519,40 @@ string_commands (int command, guint8 *p, guint8 *end, Buffer *buf)
 
 	return ERR_NONE;
 }
+static void create_file_to_check_memory_address (void) 
+{
+	if (file_check_valid_memory > 0)
+		return;
+	char *file_name = g_strdup_printf ("debugger_check_valid_memory.%d", getpid());
+	filename_check_valid_memory = g_build_filename (g_get_tmp_dir (), file_name, (const char*)NULL);
+	file_check_valid_memory = open(filename_check_valid_memory, O_CREAT | O_WRONLY | O_APPEND, S_IWUSR);
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: filename_check_valid_memory  is leaked.

File: mono/mini/debugger-agent.c [hunk 3]
```diff
@@ -9509,17 +9519,40 @@ string_commands (int command, guint8 *p, guint8 *end, Buffer *buf)
...
+		ret = FALSE;
+	}
+	return ret;
+}
 
 static ErrorCode
 pointer_commands (int command, guint8 *p, guint8 *end, Buffer *buf)
 {
+	fflush(stdout);
 	ErrorCode err;
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: Unrelated.

File: mono/mini/debugger-agent.c (hunk 2 above)
Comment (thaystg) [reply mono/mini/debugger-agent.c]: it's being deallocated on mono_debugger_agent_cleanup

File: mono/mini/debugger-agent.c [hunk 4]
```diff
@@ -10219,7 +10270,7 @@ mono_debugger_agent_init (void)
 	cbs.debug_log = debugger_agent_debug_log;
 	cbs.debug_log_is_enabled = debugger_agent_debug_log_is_enabled;
 	cbs.send_crash = mono_debugger_agent_send_crash;
-
+	
 	mini_install_dbg_callbacks (&cbs);
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: Extra whitespace.

Comment (thaystg) [reply mono/mini/debugger-agent.c]: Removed!

File: mono/mini/debugger-agent.c [hunk 5]
```diff
@@ -9506,13 +9523,42 @@ string_commands (int command, guint8 *p, guint8 *end, Buffer *buf)
...
+	g_free (file_name);
+}
+
+static gboolean 
+valid_memory_address (gpointer addr, gint size)
+{
+	gboolean ret = TRUE;
+	create_file_to_check_memory_address ();
+	if(file_check_valid_memory < 0) {
+       return TRUE;
```
Comment (lambdageek) [inline mono/mini/debugger-agent.c]:
```suggestion
		return TRUE;
```

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by lambdageek** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build Windows x64 FullAOT+LLVM

**Comment by thaystg** [comment]
@monojenkins build OS X x64 iOS SDK

**Comment by thaystg** [comment]
@monojenkins build Linux AArch64

**Pull request merged**